data/basal.go

```diff
@@ -0,0 +1,141 @@
+	ScheduleName *string          `json:"scheduleName" bson:"scheduleName" valid:"omitempty,required"`
+	Rate         *float64         `json:"rate,omitempty" bson:"rate,omitempty" valid:"omitempty,basalrate"`
+	Duration     *int             `json:"duration,omitempty" bson:"duration,omitempty" valid:"omitempty,basalduration"`
+	Insulin      *string          `json:"insulin,omitempty" bson:"insulin,omitempty" valid:"omitempty,basalinsulin"`
+	Value        *int             `json:"value,omitempty" bson:"value,omitempty" valid:"omitempty,basalvalue"`
+	Suppressed   *SuppressedBasal `json:"suppressed,omitempty" bson:"suppressed,omitempty" valid:"omitempty,required"`
+	Base         `bson:",inline"`
+}
+
+type SuppressedBasal struct {
+	Type         *string  `json:"type" bson:"type" valid:"required"`
+	DeliveryType *string  `json:"deliveryType" bson:"deliveryType" valid:"basaldeliverytype"`
+	ScheduleName *string  `json:"scheduleName" bson:"scheduleName" valid:"omitempty,required"`
+	Rate         *float64 `json:"rate" bson:"rate" valid:"omitempty,basalrate"`
+}
+
+const (
+	BasalName = "basal"
+
+	deliveryTypeField = "deliveryType"
+	scheduleNameField = "scheduleName"
+	insulinField      = "insulin"
+	valueField        = "value"
+	rateField         = "rate"
+	durationField     = "duration"
+
+	deliveryTypeTag validate.ValidationTag = "basaldeliverytype"
+	rateTag         validate.ValidationTag = "basalrate"
+	durationTag     validate.ValidationTag = "basalduration"
+	insulinTag      validate.ValidationTag = "basalinsulin"
+	valueTag        validate.ValidationTag = "basalvalue"
+
+	injectedDelivery  = "injected"
+	scheduledDelivery = "scheduled"
+	suspendDelivery   = "suspend"
+	tempDelivery      = "temp"
+
+	levemirInsulin = "levemir"
+	lantusInsulin  = "lantus"
+
+	rateValidationLowerLimit     = 0.0
+	durationValidationLowerLimit = 0
+	valueValidationLowerLimit    = 0
+)
+
+var (
+	allowedDeliveryTypes = map[string]string{injectedDelivery: injectedDelivery, scheduledDelivery: scheduledDelivery, suspendDelivery: suspendDelivery, tempDelivery: tempDelivery}
+	allowedInsulins      = map[string]string{levemirInsulin: levemirInsulin, lantusInsulin: lantusInsulin}
+
+	basalFailureReasons = validate.ErrorReasons{
+		deliveryTypeTag: fmt.Sprintf("Must be one of %s,%s,%s,%s", injectedDelivery, scheduledDelivery, suspendDelivery, tempDelivery),
+		rateTag:         fmt.Sprintf("Must be greater than %.1f", rateValidationLowerLimit),
+		durationTag:     fmt.Sprintf("Must be greater than %d", durationValidationLowerLimit),
+		valueTag:        fmt.Sprintf("Must be greater than %d", valueValidationLowerLimit),
+		insulinTag:      fmt.Sprintf("Must be one of %s,%s", levemirInsulin, lantusInsulin),
+	}
+)
+
+//BuildBasal will build a Basal record
+func BuildBasal(datum Datum, errs validate.ErrorProcessing) *Basal {
+
+	basal := &Basal{
+		ScheduleName: ToString(scheduleNameField, datum[scheduleNameField], errs),
```

Comment:
I see where you are going with this.  It does tighten up the code, but somehow it feels a little much to me.  Not that I have a better idea, though.  :)  Perhaps partly because the function is called `ToString`, but is doing far more that just converting to a string.  Also, the duplication of the field name twice seems repetitive.  Perhaps something like:

```
ScheduleName: FetchAndValidateStringField(datum, "scheduleName", errs)
```

Furthermore, what if we added the Fetch function to the TypeBuilder and passed that into the BuildBasal function rather than the errs.  For example:

```
func BuildBasal(datum Datum, builder *validate.TypeBuilder) *Basal {

	basal := &Basal{
		ScheduleName: builder.FetchAndValidateStringField(datum, "scheduleName"),
```

Not sure if that is better, but something to think about.